File: experiments/mathexpr-spec-python/mathexpr.py
"""
Mathexpr — A math expression parser and evaluator.

Pipeline: tokenize -> parse -> evaluate
Public API: calc(expression) -> float
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import List, Union


# ---------------------------------------------------------------------------
# Types
# ---------------------------------------------------------------------------

@dataclass
class Token:
    kind: str
    value: str


@dataclass
class NumberLiteral:
    type: str  # always "number"
    value: float


@dataclass
class UnaryExpr:
    type: str  # always "unary"
    op: str
    operand: AstNode


@dataclass
class BinaryExpr:
    type: str  # always "binary"
    op: str
    left: AstNode
    right: AstNode


AstNode = Union[NumberLiteral, UnaryExpr, BinaryExpr]


# ---------------------------------------------------------------------------
# Tokenizer
# ---------------------------------------------------------------------------

def tokenize(input_str: str) -> List[Token]:
    tokens: List[Token] = []
    i = 0
    length = len(input_str)

    while i < length:
        ch = input_str[i]

        # Skip whitespace
        if ch in (' ', '\t', '\n', '\r'):
            i += 1
            continue

        # Numbers
        if ch.isdigit() or ch == '.':
            start = i
            has_dot = False
            while i < length and (input_str[i].isdigit() or input_str[i] == '.'):
                if input_str[i] == '.':
                    if has_dot:
                        raise ValueError(f"Unexpected character `.`")
                    has_dot = True
                i += 1
            tokens.append(Token("number", input_str[start:i]))
            continue

        # Two-character operators
        if ch == '*' and i + 1 < length and input_str[i + 1] == '*':
            tokens.append(Token("power", "**"))
            i += 2
            continue

        # Single-character operators and parens
        single = {
            '+': 'plus',
            '-': 'minus',
            '*': 'star',
            '/': 'slash',
            '%': 'percent',
            '(': 'lparen',
            ')': 'rparen',
        }
        if ch in single:
            tokens.append(Token(single[ch], ch))
            i += 1
            continue

        raise ValueError(f"Unexpected character `{ch}` at position {i}")

    return tokens


# ---------------------------------------------------------------------------
# Parser — recursive descent with precedence climbing
# ---------------------------------------------------------------------------

class _Parser:
    def __init__(self, tokens: List[Token]):
        self.tokens = tokens
        self.pos = 0

    def peek(self) -> Token | None:
        if self.pos < len(self.tokens):
            return self.tokens[self.pos]
        return None

    def advance(self) -> Token:
        tok = self.tokens[self.pos]
        self.pos += 1
        return tok

    def expect(self, kind: str) -> Token:
        tok = self.peek()
        if tok is None:
            raise ValueError(f"Expected {kind} but reached end of input")
        if tok.kind != kind:
            raise ValueError(f"Expected {kind}")
        return self.advance()

    # Level 1: + -  (left-assoc)
    def parse_add(self) -> AstNode:
        left = self.parse_mul()
        while True:
            tok = self.peek()
            if tok is not None and tok.kind in ('plus', 'minus'):
                self.advance()
                right = self.parse_mul()
                left = BinaryExpr("binary", tok.value, left, right)
            else:
                break
        return left

    # Level 2: * / %  (left-assoc)
    def parse_mul(self) -> AstNode:
        left = self.parse_power()
        while True:
            tok = self.peek()
            if tok is not None and tok.kind in ('star', 'slash', 'percent'):
                self.advance()
                right = self.parse_power()
                left = BinaryExpr("binary", tok.value, left, right)
            else:
                break
        return left

    # Level 3: **  (right-assoc)
    def parse_power(self) -> AstNode:
        base = self.parse_unary()
        tok = self.peek()
        if tok is not None and tok.kind == 'power':
            self.advance()
            exponent = self.parse_power()  # right-recursive
            return BinaryExpr("binary", "**", base, exponent)
        return base

    # Level 4: unary -
    def parse_unary(self) -> AstNode:
        tok = self.peek()
        if tok is not None and tok.kind == 'minus':
            self.advance()
            operand = self.parse_unary()  # chain: --5
            return UnaryExpr("unary", "-", operand)
        return self.parse_atom()

    # Level 5: numbers and (expr)
    def parse_atom(self) -> AstNode:
        tok = self.peek()
        if tok is None:
            raise ValueError("Unexpected end of input")

        if tok.kind == 'number':
            self.advance()
            return NumberLiteral("number", float(tok.value))

        if tok.kind == 'lparen':
            self.advance()
            node = self.parse_add()
            self.expect('rparen')
            return node

        raise ValueError(f"Unexpected token: {tok.kind}")


def parse(tokens: List[Token]) -> AstNode:
    if not tokens:
        raise ValueError("Unexpected end of input")
    parser = _Parser(tokens)
    ast = parser.parse_add()
    if parser.pos < len(parser.tokens):
        raise ValueError("Unexpected token after expression")
    return ast


# ---------------------------------------------------------------------------
# Evaluator
# ---------------------------------------------------------------------------

def evaluate(node: AstNode) -> float:
    if isinstance(node, NumberLiteral):
        return node.value

    if isinstance(node, UnaryExpr):
        return -evaluate(node.operand)

    if isinstance(node, BinaryExpr):
        left = evaluate(node.left)
        right = evaluate(node.right)
        if node.op == '+':
            return left + right
        if node.op == '-':
            return left - right
        if node.op == '*':
            return left * right
        if node.op == '/':
            if right == 0:
                raise ValueError("Division by zero")
            return left / right
        if node.op == '%':
            if right == 0:
                raise ValueError("Modulo by zero")
            return left % right
        if node.op == '**':
            return left ** right

    raise ValueError(f"Unknown node: {node}")


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

def calc(expression: str) -> float:
    tokens = tokenize(expression)
    if not tokens:
        raise ValueError("Empty expression")
    ast = parse(tokens)
    return evaluate(ast)

File: research/experiments/mathexpr-skill-go/mathexpr.go
package mathexpr

import (
	"fmt"
	"math"
	"strconv"
	"strings"
)

// --- token-types ---

// TokenKind represents the type of a token.
type TokenKind string

const (
	TokenNumber  TokenKind = "number"
	TokenPlus    TokenKind = "plus"
	TokenMinus   TokenKind = "minus"
	TokenStar    TokenKind = "star"
	TokenSlash   TokenKind = "slash"
	TokenPercent TokenKind = "percent"
	TokenPower   TokenKind = "power"
	TokenLParen  TokenKind = "lparen"
	TokenRParen  TokenKind = "rparen"
)

// Token represents a lexical token with a kind and string value.
type Token struct {
	Kind  TokenKind
	Value string
}

// NewToken creates a new Token.
func NewToken(kind TokenKind, value string) Token {
	return Token{Kind: kind, Value: value}
}

// --- ast-types ---

// AstNode is the interface for all AST node types.
type AstNode interface {
	astNode()
}

// NumberLiteral represents a numeric value in the AST.
type NumberLiteral struct {
	Value float64
}

func (NumberLiteral) astNode() {}

// UnaryExpr represents a unary operation (e.g., negation).
type UnaryExpr struct {
	Op      string
	Operand AstNode
}

func (UnaryExpr) astNode() {}

// BinaryExpr represents a binary operation (e.g., addition).
type BinaryExpr struct {
	Op    string
	Left  AstNode
	Right AstNode
}

func (BinaryExpr) astNode() {}

// --- tokenizer ---

// Tokenize converts a math expression string into a sequence of tokens.
func Tokenize(input string) ([]Token, error) {
	tokens := []Token{}
	i := 0
	for i < len(input) {
		ch := input[i]

		// Skip whitespace
		if ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r' {
			i++
			continue
		}

		// Numbers: digits or leading dot
		if (ch >= '0' && ch <= '9') || ch == '.' {
			start := i
			hasDot := false
			for i < len(input) && ((input[i] >= '0' && input[i] <= '9') || input[i] == '.') {
				if input[i] == '.' {
					if hasDot {
						return nil, fmt.Errorf("Unexpected character '.' at position %d", i)
					}
					hasDot = true
				}
				i++
			}
			tokens = append(tokens, NewToken(TokenNumber, input[start:i]))
			continue
		}

		// ** (power) — must check before single *
		if ch == '*' && i+1 < len(input) && input[i+1] == '*' {
			tokens = append(tokens, NewToken(TokenPower, "**"))
			i += 2
			continue
		}

		// Single-character operators
		switch ch {
		case '+':
			tokens = append(tokens, NewToken(TokenPlus, "+"))
		case '-':
			tokens = append(tokens, NewToken(TokenMinus, "-"))
		case '*':
			tokens = append(tokens, NewToken(TokenStar, "*"))
		case '/':
			tokens = append(tokens, NewToken(TokenSlash, "/"))
		case '%':
			tokens = append(tokens, NewToken(TokenPercent, "%"))
		case '(':
			tokens = append(tokens, NewToken(TokenLParen, "("))
		case ')':
			tokens = append(tokens, NewToken(TokenRParen, ")"))
		default:
			return nil, fmt.Errorf("Unexpected character '%c' at position %d", ch, i)
		}
		i++
	}
	return tokens, nil
}

// --- parser ---

type parser struct {
	tokens []Token
	pos    int
}

func (p *parser) peek() *Token {
	if p.pos >= len(p.tokens) {
		return nil
	}
	return &p.tokens[p.pos]
}

func (p *parser) advance() Token {
	t := p.tokens[p.pos]
	p.pos++
	return t
}

func (p *parser) expect(kind TokenKind) (Token, error) {
	tok := p.peek()
	if tok == nil {
		return Token{}, fmt.Errorf("Expected %s but reached end of input", kind)
	}
	if tok.Kind != kind {
		return Token{}, fmt.Errorf("Expected %s but got %s:\"%s\"", kind, tok.Kind, tok.Value)
	}
	return p.advance(), nil
}

// parseAddSub handles + and - (precedence level 1, left-associative).
func (p *parser) parseAddSub() (AstNode, error) {
	left, err := p.parseMulDiv()
	if err != nil {
		return nil, err
	}
	for {
		tok := p.peek()
		if tok == nil || (tok.Kind != TokenPlus && tok.Kind != TokenMinus) {
			break
		}
		op := p.advance()
		right, err := p.parseMulDiv()
		if err != nil {
			return nil, err
		}
		left = BinaryExpr{Op: op.Value, Left: left, Right: right}
	}
	return left, nil
}

// parseMulDiv handles *, /, % (precedence level 2, left-associative).
func (p *parser) parseMulDiv() (AstNode, error) {
	left, err := p.parsePower()
	if err != nil {
		return nil, err
	}
	for {
		tok := p.peek()
		if tok == nil || (tok.Kind != TokenStar && tok.Kind != TokenSlash && tok.Kind != TokenPercent) {
			break
		}
		op := p.advance()
		right, err := p.parsePower()
		if err != nil {
			return nil, err
		}
		left = BinaryExpr{Op: op.Value, Left: left, Right: right}
	}
	return left, nil
}

// parsePower handles ** (precedence level 3, right-associative).
func (p *parser) parsePower() (AstNode, error) {
	base, err := p.parseUnary()
	if err != nil {
		return nil, err
	}
	tok := p.peek()
	if tok != nil && tok.Kind == TokenPower {
		p.advance()
		// Right-associative: recurse into parsePower (same level)
		exp, err := p.parsePower()
		if err != nil {
			return nil, err
		}
		return BinaryExpr{Op: "**", Left: base, Right: exp}, nil
	}
	return base, nil
}

// parseUnary handles unary minus (precedence level 4).
func (p *parser) parseUnary() (AstNode, error) {
	tok := p.peek()
	if tok != nil && tok.Kind == TokenMinus {
		p.advance()
		operand, err := p.parseUnary()
		if err != nil {
			return nil, err
		}
		return UnaryExpr{Op: "-", Operand: operand}, nil
	}
	return p.parseAtom()
}

// parseAtom handles numbers and parenthesized expressions (precedence level 5).
func (p *parser) parseAtom() (AstNode, error) {
	tok := p.peek()
	if tok == nil {
		return nil, fmt.Errorf("Unexpected end of input")
	}
	switch tok.Kind {
	case TokenNumber:
		t := p.advance()
		val, err := strconv.ParseFloat(t.Value, 64)
		if err != nil {
			return nil, fmt.Errorf("Invalid number: %s", t.Value)
		}
		return NumberLiteral{Value: val}, nil
	case TokenLParen:
		p.advance() // consume '('
		expr, err := p.parseAddSub()
		if err != nil {
			return nil, err
		}
		_, err = p.expect(TokenRParen)
		if err != nil {
			return nil, fmt.Errorf("Expected rparen")
		}
		return expr, nil
	default:
		return nil, fmt.Errorf("Unexpected token %s:\"%s\"", tok.Kind, tok.Value)
	}
}

// Parse converts a slice of tokens into an AST.
func Parse(tokens []Token) (AstNode, error) {
	if len(tokens) == 0 {
		return nil, fmt.Errorf("Unexpected end of input")
	}
	p := &parser{tokens: tokens, pos: 0}
	node, err := p.parseAddSub()
	if err != nil {
		return nil, err
	}
	if p.pos < len(p.tokens) {
		return nil, fmt.Errorf("Unexpected token after expression")
	}
	return node, nil
}

// --- evaluator ---

// Evaluate walks an AST and computes the numeric result.
func Evaluate(node AstNode) (float64, error) {
	switch n := node.(type) {
	case NumberLiteral:
		return n.Value, nil
	case UnaryExpr:
		operand, err := Evaluate(n.Operand)
		if err != nil {
			return 0, err
		}
		return -operand, nil
	case BinaryExpr:
		left, err := Evaluate(n.Left)
		if err != nil {
			return 0, err
		}
		right, err := Evaluate(n.Right)
		if err != nil {
			return 0, err
		}
		switch n.Op {
		case "+":
			return left + right, nil
		case "-":
			return left - right, nil
		case "*":
			return left * right, nil
		case "/":
			if right == 0 {
				return 0, fmt.Errorf("Division by zero")
			}
			return left / right, nil
		case "%":
			if right == 0 {
				return 0, fmt.Errorf("Modulo by zero")
			}
			return math.Mod(left, right), nil
		case "**":
			return math.Pow(left, right), nil
		default:
			return 0, fmt.Errorf("Unknown operator: %s", n.Op)
		}
	default:
		return 0, fmt.Errorf("Unknown AST node type")
	}
}

// --- evaluate (root: public API) ---

// Calc evaluates a math expression string and returns the numeric result.
func Calc(expression string) (float64, error) {
	trimmed := strings.TrimSpace(expression)
	if trimmed == "" {
		return 0, fmt.Errorf("Empty expression")
	}

	tokens, err := Tokenize(trimmed)
	if err != nil {
		return 0, err
	}

	ast, err := Parse(tokens)
	if err != nil {
		return 0, err
	}

	result, err := Evaluate(ast)
	if err != nil {
		return 0, err
	}

	return result, nil
}
